File: Chapter3/ros2_ws/src/avbook_cpp_nodes/src/sobel_filter_node.cpp
#include <functional>
#include <string>

#include <cv_bridge/cv_bridge.h>
#include <opencv2/imgproc.hpp>
#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/image.hpp>

class SobelFilterNode final : public rclcpp::Node
{
public:
  SobelFilterNode() : rclcpp::Node("sobel_filter_node")
  {
    input_topic_ = this->declare_parameter<std::string>("input_topic", "/image_raw");
    output_topic_ = this->declare_parameter<std::string>("output_topic", "/image_sobel");
    ksize_ = this->declare_parameter<int>("ksize", 3);
    if (ksize_ != 1 && ksize_ != 3 && ksize_ != 5 && ksize_ != 7) {
      RCLCPP_WARN(this->get_logger(), "ksize must be 1/3/5/7; using 3");
      ksize_ = 3;
    }

    pub_ = this->create_publisher<sensor_msgs::msg::Image>(output_topic_, rclcpp::QoS(10));
    sub_ = this->create_subscription<sensor_msgs::msg::Image>(
      input_topic_, rclcpp::QoS(10),
      std::bind(&SobelFilterNode::on_image, this, std::placeholders::_1));

    RCLCPP_INFO(this->get_logger(), "Sobel %s -> %s (ksize=%d)", input_topic_.c_str(), output_topic_.c_str(), ksize_);
  }

private:
  void on_image(const sensor_msgs::msg::Image::ConstSharedPtr msg)
  {
    cv_bridge::CvImageConstPtr cv_ptr;
    try {
      cv_ptr = cv_bridge::toCvShare(msg);
    } catch (const cv_bridge::Exception & e) {
      RCLCPP_WARN(this->get_logger(), "cv_bridge failed: %s", e.what());
      return;
    }

    cv::Mat gray;
    if (cv_ptr->image.channels() == 1) {
      gray = cv_ptr->image;
    } else {
      cv::cvtColor(cv_ptr->image, gray, cv::COLOR_BGR2GRAY);
    }

    cv::Mat grad_x;
    cv::Mat grad_y;
    cv::Sobel(gray, grad_x, CV_16S, 1, 0, ksize_);
    cv::Sobel(gray, grad_y, CV_16S, 0, 1, ksize_);

    cv::Mat abs_x;
    cv::Mat abs_y;
    cv::convertScaleAbs(grad_x, abs_x);
    cv::convertScaleAbs(grad_y, abs_y);

    cv::Mat edges;
    cv::addWeighted(abs_x, 0.5, abs_y, 0.5, 0.0, edges);

    cv_bridge::CvImage out;
    out.header = msg->header;
    out.encoding = "mono8";
    out.image = edges;
    pub_->publish(*out.toImageMsg());
  }

  std::string input_topic_;
  std::string output_topic_;
  int ksize_{3};

  rclcpp::Subscription<sensor_msgs::msg::Image>::SharedPtr sub_;
  rclcpp::Publisher<sensor_msgs::msg::Image>::SharedPtr pub_;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<SobelFilterNode>());
  rclcpp::shutdown();
  return 0;
}

File: Chapter3/ros2_ws/src/avbook_py_nodes/avbook_py_nodes/sobel_filter_node.py
import cv2
import rclpy
from cv_bridge import CvBridge
from rclpy.executors import ExternalShutdownException
from rclpy.node import Node
from sensor_msgs.msg import Image


class SobelFilterNode(Node):
    def __init__(self) -> None:
        super().__init__("sobel_filter_node")

        self._input_topic = self.declare_parameter("input_topic", "/image_raw").value
        self._output_topic = self.declare_parameter("output_topic", "/image_sobel").value
        self._ksize = int(self.declare_parameter("ksize", 3).value)
        if self._ksize not in (1, 3, 5, 7):
            self.get_logger().warn("ksize must be 1/3/5/7; using 3")
            self._ksize = 3

        self._bridge = CvBridge()
        self._pub = self.create_publisher(Image, self._output_topic, 10)
        self._sub = self.create_subscription(Image, self._input_topic, self._on_image, 10)

        self.get_logger().info(f"Sobel {self._input_topic} -> {self._output_topic} (ksize={self._ksize})")

    def _on_image(self, msg: Image) -> None:
        img = self._bridge.imgmsg_to_cv2(msg, desired_encoding="passthrough")
        if img.ndim == 3:
            gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
        else:
            gray = img

        grad_x = cv2.Sobel(gray, cv2.CV_16S, 1, 0, ksize=self._ksize)
        grad_y = cv2.Sobel(gray, cv2.CV_16S, 0, 1, ksize=self._ksize)

        abs_x = cv2.convertScaleAbs(grad_x)
        abs_y = cv2.convertScaleAbs(grad_y)
        edges = cv2.addWeighted(abs_x, 0.5, abs_y, 0.5, 0.0)

        out = self._bridge.cv2_to_imgmsg(edges, encoding="mono8")
        out.header = msg.header
        self._pub.publish(out)


def main() -> None:
    rclpy.init()
    node = SobelFilterNode()
    try:
        rclpy.spin(node)
    except (KeyboardInterrupt, ExternalShutdownException):
        pass
    finally:
        node.destroy_node()
        rclpy.try_shutdown()


if __name__ == "__main__":
    main()
